include/seqan3/range/view/single_pass_input.hpp

```diff
@@ -0,0 +1,402 @@
+     */
+    //!\brief Default constructor.
+    single_pass_input_sentinel() = default;
+    //!\brief Copy constructor.
+    constexpr single_pass_input_sentinel(single_pass_input_sentinel const &) = default;
+    //!\brief Move constructor.
+    constexpr single_pass_input_sentinel(single_pass_input_sentinel &&) = default;
+    //!\brief Copy assignment.
+    constexpr single_pass_input_sentinel & operator=(single_pass_input_sentinel const &) = default;
+    //!\brief Move assignment.
+    constexpr single_pass_input_sentinel & operator=(single_pass_input_sentinel &&) = default;
+    //!\brief Destructor.
+    ~single_pass_input_sentinel() = default;
+
+    //!\brief Construction from underlying sentinel.
+    single_pass_input_sentinel(sentinel_t _s) : sentinel{_s}
+    {}
+    //!\}
+
+    /*!\name Comparison
+     * \{
+     */
+    //!\brief Comparison with iterator type.
+    template <typename view_t>
+    bool operator==(single_pass_input_iterator<view_t> const & rhs) const noexcept
+    {
+        return sentinel == rhs.cached();
+    }
+
+    //!\brief Comparison with iterator type.
+    template <typename view_t>
+    bool operator!=(single_pass_input_iterator<view_t> const & rhs) const noexcept
+    {
+        return !(*this == rhs);
+    }
+    //!/}
+};
+} // seqan3::detail
+
+//-----------------------------------------------------------------------------
+// Implementation of single pass input view.
+//-----------------------------------------------------------------------------
+
+namespace seqan3::detail
+{
+
+/*!\brief Adds single_pass_input behavior to the underlying range.
+ * \tparam urng_t The underlying range type.
+ * \implements input_range_concept
+ * \ingroup view
+ */
+template <input_range_concept urng_t>
+class single_pass_input_view : public detail::view_base
+{
+private:
+
+    //!\brief Friend declaration for seqan3::detail::single_pass_input_iterator.
+    template <typename view_t>
+    friend class single_pass_input_iterator;
+
+    //!\brief The pure range type without any reference type.
+    using pure_range_t       = std::remove_reference_t<urng_t>;
+    //!\brief The iterator type for the underlying range.
+    using urng_iterator_type = iterator_t<pure_range_t>;
+    //!\brief The sentinel type for the underlying range.
+    using sentinel           = single_pass_input_sentinel<sentinel_t<pure_range_t>>;
+
+    //!\brief The internal state.
+    struct state
+    {
+        //!\brief The underlying range.
+        urng_t             urng;
+        //!\brief The cached iterator of the underlying range.
+        urng_iterator_type cached_urng_iter{};
+    };
+
+    //!\brief Shared pointer of the data model.
+    std::shared_ptr<state> state_ptr{};
+
+public:
+    /*!\name Member types
+     * \{
+     */
+    //!\brief Always returns immutable reference type, since single_pass_input cannot change the underlying values.
+    using reference         = std::add_const_t<reference_t<remove_cvref_t<urng_t>>>;
```

Comment:
in any case, this solution is not ideal, I think what you would want to do is:
```cpp
using reference         = const_reference_t<urng_t>;
```
The difference is subtle but important, let's say urng_t has the following member types:
```cpp
using value_type = uint64_t;
using reference = foobar_proxy_type; // that allows assigning uint64_t
using const_reference = uint64_t;
```
(this makes a lot of sense, btw)

In your implementation you would define your `reference` to be a const version of the proxy, but you want it to be just the value.

Reply:
good spot